internal/extproc/translator/openai_gcpanthropic_stream.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	mut := &extprocv3.BodyMutation_Body{Body: nil}
+	for {
+		eventBlock, remaining, found := bytes.Cut(p.buffer.Bytes(), []byte("\n\n"))
+		if !found {
+			break
+		}
+
+		if err := p.writeChunk(string(eventBlock), &mut.Body); err != nil {
+			return nil, nil, LLMTokenUsage{}, err
+		}
+
+		p.buffer.Reset()
+		p.buffer.Write(remaining)
+	}
+
+	if endOfStream && p.buffer.Len() > 0 {
+		finalEventBlock := p.buffer.String()
+		p.buffer.Reset()
+
+		if err := p.writeChunk(finalEventBlock, &mut.Body); err != nil {
+			return nil, nil, LLMTokenUsage{}, err
+		}
+	}
+
+	if endOfStream {
+		p.tokenUsage.TotalTokens = p.tokenUsage.InputTokens + p.tokenUsage.OutputTokens
+		finalChunk := openai.ChatCompletionResponseChunk{
+			Object:  "chat.completion.chunk",
+			Choices: []openai.ChatCompletionResponseChunkChoice{},
+			Usage: &openai.ChatCompletionResponseUsage{
+				PromptTokens:     int(p.tokenUsage.InputTokens),
+				CompletionTokens: int(p.tokenUsage.OutputTokens),
+				TotalTokens:      int(p.tokenUsage.TotalTokens),
+			},
+		}
+
+		// Add active tool calls to the final chunk.
+		var toolCalls []openai.ChatCompletionMessageToolCallParam
+		for _, tool := range p.activeToolCalls {
+			toolCalls = append(toolCalls, openai.ChatCompletionMessageToolCallParam{
+				ID:   &tool.id,
+				Type: openai.ChatCompletionMessageToolCallTypeFunction,
+				Function: openai.ChatCompletionMessageToolCallFunctionParam{
+					Name:      tool.name,
+					Arguments: tool.inputJSON,
+				},
+			})
+		}
+
+		if len(toolCalls) > 0 {
+			delta := openai.ChatCompletionResponseChunkChoiceDelta{
+				ToolCalls: toolCalls,
+			}
+			finalChunk.Choices = append(finalChunk.Choices, openai.ChatCompletionResponseChunkChoice{
+				Delta: &delta,
+			})
+		}
+
+		if finalChunk.Usage.PromptTokens > 0 || finalChunk.Usage.CompletionTokens > 0 || len(finalChunk.Choices) > 0 {
+			chunkBytes, err := json.Marshal(finalChunk)
+			if err != nil {
+				return nil, nil, LLMTokenUsage{}, fmt.Errorf("failed to marshal final stream chunk: %w", err)
+			}
+			// Write the final chunk to the response body.
+			mut.Body = append(mut.Body, sseDataPrefix...)
+			mut.Body = append(mut.Body, chunkBytes...)
+			mut.Body = append(mut.Body, '\n', '\n')
+		}
+		// Add the final [DONE] message to indicate the end of the stream.
+		mut.Body = append(mut.Body, sseDataPrefix...)
+		mut.Body = append(mut.Body, sseDoneMessage...)
+		mut.Body = append(mut.Body, '\n', '\n')
+	}
+
+	return &extprocv3.HeaderMutation{}, &extprocv3.BodyMutation{Mutation: mut}, p.tokenUsage, nil
+}
+
+func (p *anthropicStreamParser) parseAndHandleEvent(eventBlock string) (*openai.ChatCompletionResponseChunk, error) {
+	var eventType string
+	var eventData []byte
+
+	lines := strings.Split(eventBlock, "\n")
+	for _, line := range lines {
+		// Per the SSE spec, lines starting with a colon are comments and should be ignored.
+		if strings.HasPrefix(line, ":") {
+			continue
+		}
```

Comment:
i think we can remove this check as a few lines below we only deal with the matching lines anyways?

Reply:
good catch